backend/src/map_model.rs

```diff
@@ -567,34 +622,94 @@ impl MapModel {
         Ok(())
     }
 
+    pub fn router_input_before(&self) -> impl RouterInput + use<'_> {
+        struct RouterInputBefore<'a> {
```

Comment:
This is a cool pattern to get around the partial-borrow-of-some-struct-fields problem!

Reply:
(And to handle different input road lists, as you pointed out in another comment)